New ansible module for aws Redshift and Redshift Subnet Group

j-carl opened the pull request:
Hey,

I coded two modules to handle AWS Redshift cluster and Redshift Subnet Group:

- module `redshift` handles creating, deleting, modifying, and facts gathering of Redshift clusters
- module `redshift_subnet_group` handles creating, modifying, and deleting of Redshift Subnet Groups

I used the modules `rds` and `rds_subnet_group` as templates to create these modules.

Jens

gregdek commented:
@j-carl Thanks for submitting this new module.

To help facilitate reviews, we’ve broadened the number of people who can approve modules for inclusion into the Extras repository. The list of official reviewers can be found here:

https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md 

Our new policy is that if a new module is reviewed and approved by at least two official module reviewers, and receives no -1 votes, the module will be approved for inclusion.  We will be asking the community of reviewers to take a look at these modules on a regular basis.

To ensure that your module has the best chance of being approved, please double-check that you adhere to the Ansible module guidelines: http://docs.ansible.com/developing_modules.html#module-checklist

robynbergeron commented:
Adding new process. We will be evaluating all new module PRs according to this process, effective immediately. 

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

robynbergeron commented:
Pinging AWS crowd to solicit reviews hopefully: :) 

@jsmartin @scicoin-project @tombamford @garethr @jarv @jsdalton @silviud @adq @zbal @zeekin @willthames @lwade @carsongee @defionscode
@tastychutney @bpennypacker @loia

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,521 @@
+    changed = True
+    # Package up the optional parameters
+    params = {}
+    for p in ( 'db_name', 'cluster_type', 'cluster_security_groups', 'vpc_security_group_ids', 'cluster_subnet_group_name', 'availability_zone', 'preferred_maintenance_window', 'cluster_parameter_group_name', 'automated_snapshot_retention_period', 'port', 'cluster_version', 'allow_version_upgrade', 'number_of_nodes', 'publicly_accessible', 'encrypted', 'elastic_ip' ):
```
I think this line could benefit from a bit of wrapping (arrays are pretty easy to split over lines after all)

herbyg-axial commented:
@j-carl , I added some improvements to this in #1365 - I'd like to know how best to proceed

kylelundstedt commented:
@robynbergeron Any word on whether the Redshift-related pull requests will be accepted soon?  We'd love to see these features added to Ansible!  Thanks!

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,532 @@
+    description:
+      - Specifies the action to take.
+    required: true
+    default: null
```
From http://docs.ansible.com/ansible/developing_modules.html#module-checklist
default is not needed for required: true

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,532 @@
+  username:
+    description:
+      - Master database username. Used only when command=create.
+    required: true
```
If it's only used when command=create, it's not required - this does not match the module argument_spec

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,532 @@
+
+
+def describe_cluster(module, redshift):
+    """
```
Not sure there's much point having an empty docstring

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,532 @@
+  node_type:
+    description:
+      - The node type of the cluster. Must be specified when command=create.
+    required: true
```
This does not match the module argument_spec

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,532 @@
+    )
+
+    if not HAS_BOTO:
+        module.fail_json(msg='boto v2.9.0+ required for this module')
```
Not really worth specifying version here unless you're going to actually check version

willthames replied:
I added a comment here about checking version, but `from boto import redshift` does that

willthames commented on cloud/amazon/redshift_subnet_group.py:
```diff
@@ -0,0 +1,168 @@
+    required: false
+    default: null
+    aliases: ['subnets']
+  aws_secret_key:
```
This is not needed - it will come from extends_documentation_fragment: aws

willthames commented on cloud/amazon/redshift_subnet_group.py:
```diff
@@ -0,0 +1,168 @@
+    required: false
+    default: null
+    aliases: [ 'ec2_secret_key', 'secret_key' ]
+  aws_access_key:
```
This is not needed - it will come from extends_documentation_fragment: aws

gregdek commented:
Thanks @j-carl for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

j-carl commented:
Made some changes and it's ready_for_review.

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,526 @@
+'''
+
+RETURN = '''
+changed:
```
I assume changed isn't needed to be documented as it's a standard Ansible return value.

Thanks for adding the rest though

willthames commented on cloud/amazon/redshift_subnet_group.py:
```diff
@@ -0,0 +1,165 @@
+'''
+
+RETURN = '''
+changed:
```
Is there no subnet group information returned? (as with the other module, I don't think changed needs to be documented). I'd say you're missing a trick if you're not returning anything (as things like subnet id are useful to be passed to later modules in the play)

willthames commented on cloud/amazon/redshift_subnet_group.py:
```diff
@@ -0,0 +1,165 @@
+            changed = True
+
+    except boto.exception.JSONResponseError, e:
+        # FIXME: Change this to just set the error message when
```
delete the FIXMEs

willthames commented:
Only really two remaining areas of concern - the amount of information that redshift_subnet_group returns could be vastly improved, and the FIXME comments should all go (as they won't be fixed in boto, so they'll never get fixed here). Once that's done this will meet guidelines. However, once this meets guidelines, I suspect we'll want to squash commits - so you might prefer to do that before the next revision round.

needs_revision

willthames commented:
There is an argument for the facts stuff to be split into another module, and also an argument that this module should use state=present/absent. Given this module's provenance, (it's based on rds which has the same flaws), I'd just say we accept this once the remaining wrinkles are gone so that people can actually use redshift, and let the perfect implementation come later.

gregdek commented:
Thanks @j-carl for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

j-carl commented:
@willthames I didn't add any information to the redshift_subnet_group module, because the underlying function doesn't expose any id or so. If you want to use the group in other Redshift things you the group-name, which you already specified in the module (parameter: name). But I can add it if you really want.

j-carl commented:
@willthames On second thought I added some information for the module redshift_subnet_group and squashed my commits into one. So again it's ready_for_review.

linuxdynasty commented:
@j-carl I was about to write my own module until I found yours. Seems to work as expected, but shouldn't all new AWS modules be written using Boto3 instead of boto?

j-carl commented:
@linuxdynasty you right, but I wrote these modules about one year ago. I think at that point boto3 wasn't fully supported within ansible.

bryanhiestand commented:
I'm not sure why this is still in community_review after nearly two years, but this has been working for me for a while now.

kylelundstedt commented:
Ditto @bryanhiestand; love to see these Redshift modules moved out of community_review!

juliedavila commented:
@linuxdynasty this isn't considered "new" as it pertains to the boto3 only rule for new submissions. This was submitted well before the boto3 mandate. I would be okay with this being merged in. cc/ @ryansb

ryansb commented:
Going to consider the comments above about folks using this module as-is as shipits. shipit

gregdek commented:
Thanks again to @j-carl for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented:
Merged, thanks.
